core/arch/arm/include/barrier.h

```diff
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+
+#ifndef __BARRIER_H
+#define __BARRIER_H
+
+#include <compiler.h>
+#include <arm.h>
+
+#define barrier_mem()		dsb()
+#define barrier_write()		dsb_ishst()
+#define barrier_read()		dsb_ishld()
```

Comment:
This commit raised a lot of interrogations and I have spent some time reading the Arm ARM as well as the Linux kernel sources... here are my questions:

1. Wouldn't it be easier to use the same names as the Linux kernel? See [`mb(), rmb(), wmb()`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/memory-barriers.txt?h=v5.17-rc3#n1816).
2. Is "inner shareable" the correct domain here? I would rather expect "full system" shareability for such generic barriers, that is `dsb_st()` and `dsb_ld()`. Related to this: before this series, `core_mmu_map_pages()` uses `dsb_ishst()` which you replace with `barrier_write()`. Is "inner shareable" correct in this context? Shouldn't it be at least "outer shareable" so that cores from a different cluster can see the update?
3. Wouldn't a `dmb` be good enough, rather than a `dsb`?
4. Finally, I notice in the Linux kernel doc that the memory barrier macros imply a compiler barrier, and their implementations indeed have a `"memory"` constraint (see [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/include/asm/barrier.h?h=v5.17-rc3#n22) for arm64 for example). Shouldn't we update our definitions of `dmb()` and `dsb()` accordingly?

Reply:
1. Yes, I considered that. The problem is that they are not exactly the same thing, more on that below.
4. Good point, I'll fix.

How about we drop these barriers and provide a special `core_mmu_table_write_barrier()` in `core_mmu_private.h` only to be used for this special purpose?

We wait with generic barriers outside architecture specific code until we have a better use case. Or the solution is perhaps always to provide special purpose barriers? We'll see.

For most cases we have Arm recommendations on which barrier to use, to try to cover this with generic barriers as a glue layer might be too much for us to handle.

Reply:
I agree, let's do the minimum we need at this point, i.e., `core_mmu_table_write_barrier()` sounds good to me.